File: C/27_ShortestDistanceBacktracking/BOJ_02_14002_가장긴증가하는부분수열4.c
#include <stdio.h>
#include <stdlib.h>
#define math_max(a, b) a > b ? a : b
int stack[1001], idx = -1;
void push(int value) {
	stack[++idx] = value;
}
int pop() {
	if (idx < 0) return -1;
	else return stack[idx--];
}
int is_empty() {
	if (idx < 0) return 1;
	else return 0;
}
main() {
	int n;
	scanf("%d", &n);
	int* arr = (int*)malloc(sizeof(int) * n);
	int* dp = (int*)malloc(sizeof(int) * n);
	for (int i = 0; i < n; i++) {
		scanf("%d", &arr[i]);
		dp[i] = 1;
	}
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < i; j++) {
			if (arr[i] > arr[j] && dp[i] < dp[j] + 1) {
				dp[i] = dp[j] + 1;
			}
		}
	}
	int max = 0;
	for (int i = 0; i < n; i++) {
		max = math_max(max, dp[i]);
	}
	printf("%d\n", max);
	while (max > 0) {
		n--;
		if (max == dp[n]) {
			push(arr[n]);
			max--;
		}
	}
	while (!is_empty()) {
		printf("%d ", pop());
	}
}

File: JAVA/27_ShortestDistanceBacktracking/BOJ_02_14002_가장긴증가하는부분수열4.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;
import java.util.Stack;

public class Main {

    static int[] arr, dp;

    public static int LIS(int idx) {
        if (dp[idx] == 0) {
            dp[idx] = 1;
            for (int i = idx - 1; i > 0; i--) {
                if (arr[i] < arr[idx]) {
                    dp[idx] = Math.max(dp[idx], LIS(i) + 1);
                }
            }
        }
        return dp[idx];
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        arr = new int[n + 1];
        dp = new int[n + 1];
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        for (int i = 1; i <= n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }
        int max = 0;
        for (int i = 1; i <= n; i++) {
            LIS(i);
            max = Math.max(max, dp[i]);
        }
        StringBuilder sb = new StringBuilder();
        sb.append(max + "\n");
        Stack<Integer> stack = new Stack<>();
        while (max > 0) {
            if (max == dp[n]) {
                stack.push(arr[n--]);
                max--;
            } else n--;
        }
        while (!stack.isEmpty()) {
            sb.append(stack.pop() + " ");
        }
        System.out.println(sb);
    }
}